jenkins/check_projects.py

```diff
@@ -39,190 +41,393 @@
-            '--member=%s' % member,
-            project,
-        ],
-        stdout=null,
-    )
-
-  if not err:
-    with lock:
-      print >>sys.stderr, 'Added %s as %s to %s' % (member, role, project)
-      updated.add(project)
+class Checker(object):
+  def __init__(self, config):
+    self.config = config
+    self.rlExec = RateLimitedExec()
+    self.results = Results()
+
+  def run(self, filt, fix=False):
+    """
+    Checks projects for correct settings.
+    """
+    def check(project, fix):
+      self.results.reportProject(project)
+
+      properties = [
+        IAMProperty(self.rlExec, self.results),
+        EnableVmZonalDNS(self.rlExec, self.results)]
+      for prop in properties:
+        logger.info('Checking project %s for %s', project, prop.name())
+        prop.checkAndMaybeUpdate(self.config, project, fix)
+
+    projects = self._loadProjects('%s/../jobs' % os.path.dirname(__file__), filt)
+    logger.info('Checking %d projects', len(projects))
+
+    runThreadsToCompletion(
+      [threading.Thread(target=check, args=(project, fix))
+      for project in sorted(projects)])
+
+    self.logSummary()
+
+  def logSummary(self):
+    logger.info('====')
+    logger.info('Summary: %d projects, %d have been updated, %d have problems',
+        *self.results.counts())
+    logger.info('====')
+
+    for key in self.results.errors:
+      project = self.results.projects[key]
+      logger.info('Project %s needs to fix: %s', key, ','.join(project.errors))
+
+    logger.info('Helpers:')
+    fixers = collections.defaultdict(int)
+    unk = ['user:unknown']
+    for project in self.results.errors:
+      helpers = self.results.projects[project].helpers
+      if not helpers:
+        helpers = unk
+      for name in helpers:
+        fixers[name] += 1
+      logger.info('  %s: %s', project, ','.join(self._sane(s) for s in sorted(helpers)))
+
+    for name, count in sorted(fixers.items(), key=lambda i: i[1]):
+      logger.info('  %s: %s', count, self._sane(name))
+
+    if self.results.counts()[2] != 0:
+      sys.exit(1)
+
+  @staticmethod
+  def _sane(member):
+    if ':' not in member:
+      raise ValueError(member)
+    email = member.split(':')[1]
+    return email.split('@')[0]
+
+  def _loadProjects(self, configs, filt):
+    """
+    Scans the project directories for GCP projects to check.
+    """
+    projects = set()
+    regex = re.compile(filt)
+    for dirname, _, files in os.walk(configs):
+      for path in files:
+        full_path = os.path.join(dirname, path)
+        logger.debug('Path = %s', path)
+        if not regex.match(path):
+          continue
+        with open(full_path) as fp:
+          for project in re.findall(
+              '^PROJECT="?([^"\r\n]+)"?$', fp.read(), re.MULTILINE):
+            if '{' not in project:
+              projects.add(project)
+              continue
+            else:
+              raise ValueError(project)
+    return projects
+
+
+class ProjectProperty(object):
```

Comment:
This kind of class is pretty pointless in python with its duck typing, can you just get rid of it?